presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ParametricScalarImplementation.java

```diff
@@ -416,38 +493,45 @@ else if (actualReturnType.isPrimitive()) {
                         "Expected type parameter to only contain A-Z and 0-9 (starting with A-Z), but got %s on method [%s]", typeParameter.value(), method);
             }
 
-            inferSpecialization(method, actualReturnType, returnType.value(), nullable);
+            inferSpecialization(method, actualReturnType, returnType.value());
             parseArguments(method);
 
             this.constructorMethodHandle = getConstructor(method, constructor);
 
             this.methodHandle = getMethodHandle(method);
 
-            ParametricScalarImplementationChoice choice = new ParametricScalarImplementationChoice(nullable, argumentProperties, methodHandle, constructorMethodHandle, dependencies, constructorDependencies);
+            ParametricScalarImplementationChoice choice = new ParametricScalarImplementationChoice(nullable, hasConnectorSession, argumentProperties, methodHandle, constructorMethodHandle, dependencies, constructorDependencies);
             choices.add(choice);
         }
 
         private void parseArguments(Method method)
         {
+            boolean encounteredNonDependencyAnnotation = false;
             int i = 0;
             while (i < method.getParameterCount()) {
                 Parameter parameter = method.getParameters()[i];
                 Class<?> parameterType = parameter.getType();
 
                 // Skip injected parameters
                 if (parameterType == ConnectorSession.class) {
+                    checkCondition(!hasConnectorSession, FUNCTION_IMPLEMENTATION_ERROR, "Method [%s] has more than 1 ConnectorSession in the parameter list", method);
                     i++;
+                    hasConnectorSession = true;
                     continue;
                 }
 
                 Optional<Annotation> implementationDependency = getImplementationDependencyAnnotation(parameter);
                 if (implementationDependency.isPresent()) {
+                    checkCondition(!encounteredNonDependencyAnnotation, FUNCTION_IMPLEMENTATION_ERROR, "Method [%s] has parameters annotated with Dependency annotations that appears after other parameters", method);
+
                     // check if only declared typeParameters and literalParameters are used
                     validateImplementationDependencyAnnotation(method, implementationDependency.get(), typeParameterNames, literalParameters);
                     dependencies.add(createDependency(implementationDependency.get(), literalParameters));
                     i++;
                 }
                 else {
+                    encounteredNonDependencyAnnotation = true;
```

Comment:
nit: name the variable `allowDependencyAnnotation`?

Reply:
I prefer the original name. We can talk.